be/src/olap/types.h

```diff
@@ -106,12 +141,184 @@ class TypeInfo {
     const size_t _size;
     const FieldType _field_type;
 
-    friend class TypeInfoResolver;
-    template<typename TypeTraitsClass> TypeInfo(TypeTraitsClass t);
+    friend class ScalarTypeInfoResolver;
+    template<typename TypeTraitsClass> ScalarTypeInfo(TypeTraitsClass t);
+};
+
+class ArrayTypeInfo : public TypeInfo {
+public:
+    explicit ArrayTypeInfo(TypeInfo* item_type_info)
+    : _item_type_info(item_type_info), _item_size(item_type_info->size()) {
+    }
+
+    inline bool equal(const void* left, const void* right) const override {
+        auto l_value = reinterpret_cast<const Collection*>(left);
+        auto r_value = reinterpret_cast<const Collection*>(right);
+        if (l_value->length != r_value->length) {
+            return false;
+        }
+        size_t len = l_value->length;
+
+        for (size_t i = 0; i < len; ++i){
+            if (l_value->null_signs[i]) {
```

Comment:
I think we could improve on the not-nullable column. If the array column isn't nullable, we don't need to check null value.